signing/signing.go

```diff
@@ -1,20 +1,363 @@
+	defaultBranch := "refs/heads/" + repo.GetDefaultBranch()
+	if defaultBranch != repoRef {
+		return fmt.Errorf("branch of cert isn't the repo's default branch")
+	}
+
+	// Get workflow file from cert commit SHA.
+	opts := &github.RepositoryContentGetOptions{Ref: repoSHA}
+	contents, _, _, err := client.Repositories.GetContents(ctx, ownerName, repoName, workflowPath, opts)
+	if err != nil {
+		return fmt.Errorf("error downloading workflow contents from repo: %v", err)
+	}
+
+	workflowContent, err := contents.GetContent()
+	if err != nil {
+		return fmt.Errorf("error decoding workflow contents: %v", err)
+	}
+
+	// Verify scorecard workflow.
+	err = verifyScorecardWorkflow(workflowContent)
+	if err != nil {
+		return fmt.Errorf("workflow could not be verified: %v", err)
+	}
+
+	// Save scorecard results (results.sarif, results.json, score.txt) to GCS
+	bucketURL := "gs://ossf-scorecard-results"
+	folderPath := fmt.Sprintf("%s/%s", "github.com", repoPath)
+	sarifPath := fmt.Sprintf("%s/results.sarif", folderPath)
+	jsonPath := fmt.Sprintf("%s/results.json", folderPath)
+
+	err1 = data.WriteToBlobStore(ctx, bucketURL, sarifPath, []byte(scorecardOutput.SarifOutput))
+	err2 = data.WriteToBlobStore(ctx, bucketURL, jsonPath, []byte(scorecardOutput.JsonOutput))
+	if err1 != nil || err2 != nil {
+		return fmt.Errorf(errorWritingBucket.Error()+": %v, %v", err1, err2)
+		// return fmt.Errorf("error writing to GCS bucket: %v, %v", err1, err2)
+	}
+	return nil
+}
+
+func lookupPayload(ctx context.Context, payload []byte) (repoPath, repoRef, repoSHA, workflowPath string, err error) {
+	// Get most recent Rekor entry uuid.
+	rekorClient, err := rekor.NewClient(options.DefaultRekorURL)
+	if err != nil {
+		return "", "", "", "", fmt.Errorf("error initializing Rekor client: %v", err)
+	}
+
+	uuids, err := cosign.FindTLogEntriesByPayload(ctx, rekorClient, []byte(payload))
+	if err != nil || len(uuids) == 0 {
+		return "", "", "", "", fmt.Errorf("error finding tlog entries corresponding to payload: %v", err)
+	}
+	uuid := uuids[len(uuids)-1] // ignore past entries.
+
+	// Get tlog entry from the UUID.
+	entry, err := cosign.GetTlogEntry(ctx, rekorClient, uuid)
+	if err != nil {
+		return "", "", "", "", fmt.Errorf("error fetching tlog entry: %v", err)
+	}
+
+	// Verify tlog entry to make sure it is actually in the log.
+	err = cosign.VerifyTLogEntry(ctx, rekorClient, entry)
+	if err != nil {
+		return "", "", "", "", fmt.Errorf("error verifying tlog entry: %v", err)
+	}
+
+	// Extract certificate.
+	certs, err := extractCerts(entry)
+	if err != nil || len(certs) == 0 {
+		return "", "", "", "", fmt.Errorf("error extracting certificate from entry: %v", err)
+	}
+	if len(certs) > 1 {
+		return "", "", "", "", fmt.Errorf("multiple certificates found for the entry: %v", err)
+	}
+
+	cert := certs[0]
+
+	// Verify that cert isn't expired.
+	if err = cosign.CheckExpiry(cert, time.Unix(*entry.IntegratedTime, 0)); err != nil {
```

Comment:
do you also need to verify the signed ET from rekor?

Reply:
I think this handles it because the entry is from Rekor?

Reply:
I'm not certain, but I thought we needed to verify it. I think it removes some level of trust from rekor. I may be wrong, though. @asraa can you comment?

Reply:
That happens inside VerifyTLogEntry https://github.com/sigstore/cosign/blob/80fe5a31a398c45395d09d23e237c676ddad709f/pkg/cosign/tlog.go#L387

Yes, you need to do the SET verification. Otherwise you have no guarantee that this entry UUID is in /Rekors/ log. (It could serve you any inclusion proof from any log)

Reply:
so no need to do anything here. `VerifyTLogEntry` does enough